### ISSUE TITLE
Policy issues

### ISSUE THREAD

**Issue by LennartAJansson** [Squirrel/Squirrel.Windows#843]
We have issues due to execution policy restrictions at our company. We keep a whitelist over folders where execution is ok and would like to be able to define where Update.exe is unpacking and executing. As it is right now it uses %localappdata%\SquirrelTemp with a fallback on C:\ProgramData\%username%\SquirrelTemp, would it be possible to replace SquirrelTemp with some environmentvariable or parameter?

**Comment by elmahdibenhadda** [comment]
Hi,

You can pass the root installation directory to the UpdateManager.
[https://github.com/Squirrel/Squirrel.Windows/blob/master/src/Squirrel/UpdateManager.cs#L34](url)
Hope it helps.

Mehdi.

**Comment by LennartAJansson** [comment]
Tried that:
using (var mgr = new UpdateManager("MyDeployPath",null, "AnotherPath"))
The result was that I got some folders and files in "AnotherPath" but Update.exe was totally blown away so when I try to start the application it only fail saying that it can't find that...

Lennart

**Comment by LennartAJansson** [comment]
Further investigations show:
I have a build and deploy scenario that deploys to a network path, same network path is first parameter to UpdateManager constructor. As a third parameter I have the local folder where we are allowed to execute user-level installations (Clickonce etc).

When I try to perform a clean install on a client protected by the policy it fails with a C++ error related to the Fallback path above (Installation has failed - Failed to extract installer). If I search my folders I can see that it populated %localappdata%\SquirrelTemp with my nupkg's, RELEASES and Update.exe. No trace is found in the allowed path and no log is generated enywhere.

When I try to perform a clean install of the same package on an unprotected server it installs and starts initially BUT not in the path according to the third parameter in the constructor. ALSO, the Update.exe is missing, the shortcut is pointing to the allowed path and there's no Update.exe to be found. The log file is located in the default folder %localappdata%\SquirrelTemp and it shows following in the log:

2016-10-04 09:15:08> ApplyReleasesImpl: About to save shortcut: C:\Users\JohnDoe\Desktop\MyApp.lnk (target C:\Users\JohnDoe\AppData\Local\MyApp\Update.exe, workingDir C:\Users\JohnDoe\AppData\Local\MyApp\app-16.10.4.2, args --processStart "MyApp.exe")

BUT the shortcut that is created contains following Target:
C:\SafeFolder\SquirrelTemp\MyApp\update.exe --processStart "MyApp.exe"

Nowhere in the log it says anything about the safefolder? And the initial startup of the application post install is performed against the copy that resides in %LocalAppData%\MyApp\app-16.10.4.2

Attaching the log, I will continue to find out what the heck is happening... :)

/Lennart

[SquirrelSetup.txt](https://github.com/Squirrel/Squirrel.Windows/files/507886/SquirrelSetup.txt)

**Comment by LennartAJansson** [comment]
Final investigations:

**The source to MyApp.exe contains:**

using (var mgr = new UpdateManager(@"\\MyServer\SquirrelClient\MyApp\Releases", null, @"C:\AllowedPath\SquirrelTemp"))
{
	ReleaseEntry result = await mgr.UpdateApp();
}

Our build and deploy process will copy all built files to \\MyServer\SquirrelClient\MyApp\Releases, 
this is where all future updates should be rolled out from.
In the scenarios below I have tried different ways, between the tests I did a complete cleanup of everything, shortcuts, %LocalAppData%\SquirrelTemp, %LocalAppData%\MyApp, C:\ProgramData and C:\AllowedPath\SquirrelTemp.


**Installing on a server without execution policy restrictions:**
I did this type of install just to verify what is happening...

As user User1 I execute \\MyServer\SquirrelClient\MyApp\Releases\Setup.exe.

Application installs and starts initially as:
C:\Users\User1\AppData\Local\MyApp\app-16.10.4.2\MyApp.exe.
Looking in that folder I can see a full installation WITH Update.exe.

Installation is also found under C:\AllowedPath\SquirrelTemp (according to constructor call above).
Looking in that folder I can see a full installation WITHOUT Update.exe.

The generated shortcuts says:
Target on startmenu: C:\AllowedPath\SquirrelTemp\MyApp\update.exe --processStart "MyApp.exe".
Target on desktop: C:\AllowedPath\SquirrelTemp\MyApp\update.exe --processStart "MyApp.exe".

Log file is located in: C:\Users\User1\AppData\Local\SquirrelTemp.


**Installing on a client with execution policy restrictions:**
1st scenario: As restricted User2 I execute \\MyServer\SquirrelClient\MyApp\Releases\Setup.exe.

2nd scenario: I copy \\MyServer\SquirrelClient\MyApp\Releases to C:\AllowedPath\SquirrelClient\MyApp\Releases since it's the location where I'm allowed to execute.
As restricted User2 I execute C:\AllowedPath\SquirrelClient\MyApp\Releases\Setup.exe.

In both scenarios:
Setup creates C:\Users\User2\AppData\Local\SquirrelTemp and copies the installation files there. It results in a fallback.
Setup creates C:\ProgramData\User2\SquirrelTemp (the Fallback folder according to Setup.exe) and copies the installation files there.

Setup.exe fails after trying to use both folders with error message: Installation has failed - Failed to extract installer. This is because I'm not allowed to execute anything here.

No log is generated.


**Our requirements:**
Setup.exe should unpack installation to a subfolder inside AllowedPath where it's allowed to execute and use that folder during installation.
Installation should be made completely (with Update.exe) to a subfolder inside AllowedPath.
If we look at AllowedPath it would be ok if AllowedPath is the third parameter to the constructor and the installation is executed in a Temp subfolder and installed to a MyApp subfolder under AllowedPath. Temp folder could be deleted afterwards, logfile should be saved in the MyApp folder.
AllowedPath should be possible to be sent on the commandline when executing Setup.exe.
Every attempt to execute anything inside any temp folders like %temp% is prohibited.


**Comments:**
I don't think this scenario is odd, most companies with higher degree of execution policies experience the same issues. 
We have also tried to use a product named CatLight (https://catlight.io/) and that application suffers from the same restrictions because it is deployed with Squirrel. 

If one tries to execute several Squirrel based installers for different products they will all end up in %LocalAppData%\SquirrelTemp and overwrite any Update.exe and RELEASES files that already exists. Adding the product name to the unpack subfolder structure would make it more clean.

So to make Squirrel a really powerful installer it should be capable of addressing these basic requirements. If I can be of any assistance, please let me know!

**Comment by Andrew-Hanlon** [comment]
Have you tried setting the environment variable "SQUIRREL_TEMP"? I am not sure if it is implemented in the native installer code, but Squirrel itself checks that variable first before using the special folders.

**Comment by LennartAJansson** [comment]
Hi, just tried that, no difference so most likely it's not implemented. Setup tried to extract and execute from %LocalAppData%\SquirrelTemp and C:\ProgramData\UserName\SquirrelTemp

Thanks for the hint though :)

**Comment by LennartAJansson** [comment]
This is the output from my powershell script that creates my installer, as you can see I'm using 1.4.4
VERBOSE: Using: E:\Agent1\_work\13\s\packages\squirrel.windows.1.4.4\tools\squirrel.exe
VERBOSE: Arguments:  --releasify E:\Agent1\_work\13\s\MyApp\bin\release\MyApp.16.10.04.4.nupkg

On the client I:
Startup a commandprompt
SET SQUIRREL_TEMP=C:\AllowedPath\SquirrelTemp
Executes Setup.exe
Setup.exe fails
So I can't even reach the managed part, it breaks in the native...

**Comment by LennartAJansson** [comment]
I tried this in UpdateRunner.cpp:

```
	wchar_t *envSquirrelTemp = _wgetenv(L"SQUIRREL_TEMP");
	if (envSquirrelTemp) {
		_swprintf_c(targetDir, _countof(targetDir), L"%s", envSquirrelTemp);
	}
	else {
		if (!useFallbackDir) {
			SHGetFolderPath(NULL, CSIDL_LOCAL_APPDATA, NULL, SHGFP_TYPE_CURRENT, targetDir);
		}
		else {
			wchar_t username[512];
			wchar_t uid[128];
			wchar_t appDataDir[MAX_PATH];
			ULONG unameSize = _countof(username);

			SHGetFolderPath(NULL, CSIDL_COMMON_APPDATA, NULL, SHGFP_TYPE_CURRENT, appDataDir);
			GetUserName(username, &unameSize);
			DWORD lastError = GetLastError();

			_swprintf_c(targetDir, _countof(targetDir), L"%s\\%s", appDataDir, username);

			if (!CreateDirectory(targetDir, NULL) && GetLastError() != ERROR_ALREADY_EXISTS) {
				wchar_t err[4096];
				_swprintf_c(err, _countof(err), L"Unable to write to %s - IT policies may be restricting access to this folder", targetDir);
				DisplayErrorMessage(CString(err), NULL);

				return -1;
			}
		}


		wcscat_s(targetDir, _countof(targetDir), L"\\SquirrelTemp");

	}
```
This give me the expected behaviour in Setup.exe
However, when Update.exe is executed with a specific targetfolder in the constructor there's a bug that won't copy Update.exe to that location. I will investigate that a bit further as well...

/Lennart

**Comment by Thieum** [comment]
Fixed by #846
